### DISCUSSION TITLE
What are the possible consequences of reusing the same session throughout the app?

### DISCUSSION THREAD

**Discussion by zmievsa** [encode/httpx#2719] [Q&A] [answered]
We use httpx.AsyncSession for each internal host. I.e. service "dogs" makes requests to service "cats". Hence service "dogs" defines a single session for communications with service "cats". 

We open the session at the beginning of a fastapi process and never close it. 

The entire thing happens within kubernetes. 

Would our approach with sessions result in any unwanted behavior? I failed to find any docs or other resources that would indicate that we would encounter any issues. 

Note that I understand that httpx recommends always closing the session and now I'm just trying to understand the reasons behind this ;)

**Comment by karpetrosyan** [comment]
Before anything else, it's important to understand that each connection requires file descriptors that are kept in the process memory.
**If you intend to use the session continuously**, you can leave it open until the process is finished because the kernel automatically closes all open files when a process terminates. 
But you should explicitly close the session if you decide you **no longer need them**. Otherwise, problems like the file descriptor limit being exceeded might occur.

**Comment by lovelydinosaur** [comment, marked as answer]
> We open the session at the beginning of a fastapi process and never close it.

Great! Use a single client all the way through the lifetime of your process is how you want it. You'll get proper connection pooling throughout that way.

> Note that I understand that httpx recommends always closing the session and now I'm just trying to understand the reasons behind this

I'm not sure how much it matters, you'd need to really dig into the differences between how the outstanding TCP connections are closed depending on when you *do* explicitly close the connection pool at the end of it's lifespan vs. when you just quit the Python process.
